service/history/workflow/mutable_state_impl.go

```diff
@@ -4955,33 +4962,78 @@ func (ms *MutableStateImpl) updateActivityInfoForRetries(
 	nextAttempt int32,
 	activityFailure *failurepb.Failure,
 ) {
-	ms.updateActivity(ai, func(info *persistencespb.ActivityInfo, impl *MutableStateImpl) *persistencespb.ActivityInfo {
-		ai = UpdateActivityInfoForRetries(
-			ai,
-			ms.GetCurrentVersion(),
-			nextAttempt,
-			ms.truncateRetryableActivityFailure(activityFailure),
-			timestamppb.New(nextScheduledTime),
-		)
-		return ai
+	ms.UpdateActivityWithCallback(ai, func(info *persistencespb.ActivityInfo, mutableState MutableState) {
+		mutableStateImpl, ok := mutableState.(*MutableStateImpl)
+		if ok {
+			ai = UpdateActivityInfoForRetries(
+				ai,
+				mutableStateImpl.GetCurrentVersion(),
+				nextAttempt,
+				mutableStateImpl.truncateRetryableActivityFailure(activityFailure),
+				timestamppb.New(nextScheduledTime),
+			)
+		}
 	})
 }
 
-func (ms *MutableStateImpl) updateActivity(ai *persistencespb.ActivityInfo, updateCallback func(*persistencespb.ActivityInfo, *MutableStateImpl) *persistencespb.ActivityInfo) {
-	// we need to store activity info size since pendingActivityInfoIDs holds pointers to activity
-	// info and if prev found it points to the same activity info as ai, so updating ai will cause
-	// size of prev change.
+func (ms *MutableStateImpl) UpdateActivityWithCallback(
+	ai *persistencespb.ActivityInfo,
+	updateCallback UpdateActivityCallback,
+) {
+	// incoming activity info can be a pointer to the activity info in pendingActivityInfoIDs
+	// we need to store activity info size since pendingActivityInfoIDs holds pointers to activity info
+	// if prev found it can point to the same activity info as incoming activity info,
 	var originalSize int
 	if prev, ok := ms.pendingActivityInfoIDs[ai.ScheduledEventId]; ok {
 		originalSize = prev.Size()
 	}
+
 	updateCallback(ai, ms)
 
 	ms.approximateSize += ai.Size() - originalSize
 	ms.updateActivityInfos[ai.ScheduledEventId] = ai
 	ms.syncActivityTasks[ai.ScheduledEventId] = struct{}{}
 }
 
+func (ms *MutableStateImpl) UpdatePausedEntitiesSearchAttribute() error {
+	// TODO: search attributes are coupled with mutable state.
+	// This should be refactored so we don't add new function to MutableState interface every time we want to work with search attributes.
+
+	pausedInfoMap := make(map[string]struct{})
+
+	for _, ai := range ms.GetPendingActivityInfos() {
+		if !ai.Paused {
+			continue
+		}
+		pausedInfoMap[ai.ActivityType.Name] = struct{}{}
+	}
+	pausedInfo := make([]string, 0, len(pausedInfoMap))
+	for activityType := range pausedInfoMap {
+		activityType = url.QueryEscape(activityType)
```

Comment:
You don't need to escape if the set of value types is limited and cannot contain colons. Escaping was only needed when we considered packing three values into this attribute.

Reply:
In our documentation activity type is defined as "is the mapping of a name to an Activity Definition." There is no restrictions. In probably most SDK activity type is just a function name, which of cause don't need to be escaped. But we don't guarantee that, at least I don't see it in the documentation

Reply:
ok lets remove it